src/histogram.rs

```diff
@@ -67,8 +92,142 @@ impl Drop for HistogramTimer {
+                    .chain(once(f64::MAX))
+                    .map(|upper_bound| (upper_bound, AtomicU64::new(0)))
+                    .collect(),
+            }),
+        }
+    }
+
+    pub fn observe(&self, nanos: u64) {
+        self.observe_and_bucket(nanos);
+    }
+
+    pub(crate) fn observe_and_bucket(&self, v: u64) -> Option<usize> {
+        self.inner.sum.fetch_add(v, Ordering::Relaxed);
+        self.inner.count.fetch_add(1, Ordering::Relaxed);
+
+        let first_bucket = self
+            .inner
+            .buckets
+            .iter()
+            .enumerate()
+            .find(|(_i, (upper_bound, _value))| upper_bound >= &(v as f64 * 1E-9));
+
+        match first_bucket {
+            Some((i, (_upper_bound, value))) => {
+                value.fetch_add(1, Ordering::Relaxed);
+                Some(i)
+            }
+            None => None,
+        }
+    }
+
+    pub(crate) fn get(&self) -> (f64, u64, Vec<(f64, u64)>) {
+        let sum = seconds(self.inner.sum.load(Ordering::Relaxed));
+        let count = self.inner.count.load(Ordering::Relaxed);
+        let buckets = self
+            .inner
+            .buckets
+            .iter()
+            .map(|(k, v)| (*k, v.load(Ordering::Relaxed)))
+            .collect();
+        (sum, count, buckets)
+    }
+}
+
+impl TypedMetric for TimeHistogram {
+    const TYPE: MetricType = MetricType::Histogram;
+}
+
+#[inline(always)]
+fn seconds(val: u64) -> f64 {
+    (val as f64) * 1E-9
+}
+
+fn encode_histogram_with_maybe_exemplars<S: Encode>(
+    sum: f64,
+    count: u64,
+    buckets: &[(f64, u64)],
+    exemplars: Option<&HashMap<usize, Exemplar<S, f64>>>,
+    mut encoder: Encoder,
+) -> Result<(), std::io::Error> {
+    encoder
+        .encode_suffix("sum")?
+        .no_bucket()?
+        .encode_value(sum)?
+        .no_exemplar()?;
+    encoder
+        .encode_suffix("count")?
+        .no_bucket()?
+        .encode_value(count)?
+        .no_exemplar()?;
+
+    let mut cummulative = 0;
+    for (i, (upper_bound, count)) in buckets.iter().enumerate() {
+        cummulative += count;
+        let mut bucket_encoder = encoder.encode_suffix("bucket")?;
+        let mut value_encoder = bucket_encoder.encode_bucket(*upper_bound)?;
+        let mut exemplar_encoder = value_encoder.encode_value(cummulative)?;
+
+        match exemplars.and_then(|es| es.get(&i)) {
+            Some(exemplar) => exemplar_encoder.encode_exemplar(exemplar)?,
+            None => exemplar_encoder.no_exemplar()?,
+        }
+    }
+
+    Ok(())
+}
+
+impl EncodeMetric for TimeHistogram {
+    fn encode(&self, encoder: Encoder) -> Result<(), std::io::Error> {
+        let (sum, count, buckets) = self.get();
+        // TODO: Would be better to use never type instead of `()`.
+        encode_histogram_with_maybe_exemplars::<()>(sum, count, &buckets, None, encoder)
+    }
+
+    fn metric_type(&self) -> MetricType {
+        Self::TYPE
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Please move those to `tests/histogram.rs`

Reply:
I can move these to a separate test mod, but I can't move them to an integration tests file as they rely on unexposed internals that I do not wish to expose.

Reply:
Specifically, the `get` method is an internal implementation detail.

Reply:
I'm fine with having histogram be a folder module with a dedicated tests module file.